data/campaigns/Eastern_Invasion/utils/amlas.cfg

```diff
@@ -0,0 +1,318 @@
+#textdomain wesnoth-ei
+
+#define DEFAULT_AMLA_EFFECTS
+    [effect]
+        apply_to=max_experience
+        increase=20%
+    [/effect]
+
+    [effect]
+        apply_to=hitpoints
+        heal_full=yes
+    [/effect]
+
+    [effect]
+        apply_to=status
+        remove=poisoned
+    [/effect]
+
+    [effect]
+        apply_to=status
+        remove=slowed
+    [/effect]
+
+#enddef
+#define GLOBAL__UNIVERSAL_AMLAS
+    #############################
+    # EXPLAIN AMLAS
+    #############################
+    [event]
+        name=explain_amlas
+
+        [filter_condition]
+            {VARIABLE_CONDITIONAL finished_explaining_amlas not_equals yes}
+        [/filter_condition]
+
+        [message]
+            speaker,image=narrator,wesnoth-icon.png
+            message=_"In this campaign, whenever a max-level unit advances, they can select one of several minor bonuses!"
+        [/message]
+        {VARIABLE finished_explaining_amlas yes}
+    [/event]
+
+    #############################
+    # IMPLEMENT AMLAS
+    #############################
+    [event]
+        name=pre advance
+        first_time_only=no
+
+        [filter]
+            [not]
+                id=$amla_currently_advancing_id
+            [/not]
+        [/filter]
+
+        #--------------------
+        # SKIP IF NOT MAX-LEVEL
+        #--------------------
+        {STORE_UNIT_VAR id=$unit.id advances_to amla_candidate_advances_to}
+        [if] {VARIABLE_CONDITIONAL amla_candidate_advances_to equals ""}
+            [then]
+            [/then]
+
+            [else]
+                {CLEAR_VARIABLE amla_candidate_advances_to}
+                [return]
+                [/return]
+            [/else]
+        [/if]
+        {CLEAR_VARIABLE amla_candidate_advances_to}
+
+        #--------------------
+        # AVOID RECURSION
+        #--------------------
+        # modifying a unit re-triggers "pre advance", so avoid infinite recursion
+        {VARIABLE amla_currently_advancing_id $unit.id}
+
+        #--------------------
+        # EXPLAIN AMLAS
+        #--------------------
+        [fire_event]
+            name=explain_amlas
+        [/fire_event]
+
+        #--------------------
+        # ADD AMLAS
+        #--------------------
+        [modify_unit]
+            [filter]
+                id=$unit.id
+            [/filter]
+
+            [object]
+                id=universal_amlas_object
+                [effect]
+                    apply_to=remove_advancement
+                    amlas=amla_default
+                [/effect]
+
+                #--------------------
+                # HITPOINTS
+                #--------------------
+                [effect]
```

Comment:
Is there any reason why you need one `[effect]` tag for each AMLA instead of stuffing them all into a single `[effect]`?

Reply:
Yeah, you're right, better to put them all in a single [effect].  Another case where converting from my original macros to WML tags caused ugly duplication (not that it was strictly necessary in the originals either).  I write it like this:
```
                #--------------------
                # HITPOINTS
                #--------------------
                {EFFECT new_advancement ([advancement]
                    id,description=hitpoints_increase,_"Gain +8 hitpoints."
                    strict_amla,max_times=yes,-1
                    {EFFECT hitpoints increase_total=8}
                    {DEFAULT_AMLA_EFFECTS}
                [/advancement])}
                
                #--------------------
                # MELEE DAMAGE
                #--------------------
                {EFFECT new_advancement ([advancement]
                    id,description=melee_increase1,_"Gain +4 melee damage."
                    {FILTER( {HAS_ATTACK range,number=melee,1} {NOT({HAS_ATTACK range,number=melee,2-99})} )}
                    strict_amla,max_times=yes,-1
                    {EFFECT attack increase_damage,range=4,melee}
                    {DEFAULT_AMLA_EFFECTS}
                [/advancement])}
                {EFFECT new_advancement ([advancement]
                    id,description=melee_increase2,_"Gain +2 melee damage."
                    {FILTER( {HAS_ATTACK range,number=melee,2} {NOT({HAS_ATTACK range,number=melee,3-99})} )}
                    strict_amla,max_times=yes,-1
                    {EFFECT attack increase_damage,range=2,melee}
                    {DEFAULT_AMLA_EFFECTS}
                [/advancement])}
                {EFFECT new_advancement ([advancement]
                    id,description=melee_increase3,_"Gain +1 melee damage."
                    {FILTER( {HAS_ATTACK range,number=melee,3-99} )}
                    strict_amla,max_times=yes,-1
                    {EFFECT attack increase_damage,range=1,melee}
                    {DEFAULT_AMLA_EFFECTS}
                [/advancement])}
```